Implementation of 1step and 2step analytical orbit spectra

anvalen97 opened the pull request:
Notes:
- New packages were added, some of which turned out to be of no use. The essential ones for the new implementation are CSV, DataFrames, LoopVectorization, LegendrePolynomials and NaNMath. All the rest can be discarded actually.
- weightsWebApp changes "had" to be committed, but can all be discarded (as you know).
- One big change is how reactions are dealt with. I think identifying them by just the reactants is not enough, simply because you might want to look at one product or the other (the new cross sections account for this by flipping signs properly). I know you did not like the Reaction_name parameter2, but that is what I could come up with for now (feel free to adjust this mechanic).
- DT neutron cross section was upgraded to account for emission angle as well. Aka anisotropy.
- Most comments I had for you regarding the new script, are situated in the first commented part of it. Proper documentation will be added as we go (including referencing to the papers etc).
- Just a reminder that vcone.jl and spec.py were not actually touched by me :)

Lastly, I will work on including all remaining reactions, since for now it is just DT neutrons and alphaBe gammas. That should be fairly easy!

I will add other comments if needed.

henrikjaerleblad commented on Manifest.toml:
```diff
@@ -1,8 +1,8 @@
 # This file is machine-generated - editing it directly is not advised
 
-julia_version = "1.9.0"
+julia_version = "1.10.4"
```
This is currently not possible. We have to use Julia 1.9.X at the moment, because if we upgrade then many features in the OrbitTomography.jl package cannot be used. Specifically the PolyharmonicSplines.jl package has some functions that cannot be used in Julia 1.10. This package is used by the OrbitTomography.jl package. I am currently working on a fix for this. But until we have this fix in place, we cannot use Julia version 1.10 for the OWCF, since then the OrbitTomography.jl package would stop working.

henrikjaerleblad commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+    if reaction_name in ["TD_neutrons", "DT_neutrons", "TD_alphas", "DT_alphas"]
+        mass_ratio = (reaction_name == "TD_neutrons" || reaction_name == "TD_alphas") ? cnst.mt / (cnst.md + cnst.mt) : cnst.md / (cnst.md + cnst.mt)
+        # Read the data file
+        df = CSV.read("fit_data/dtn4He_legendre_Drosg.txt", DataFrame; delim='\t', header=false)
```
Therefore, I ask that you remove the CSV.jl and DataFrames.jl packages, and use the DelimitedFiles.jl package instead :) Unless I am misunderstanding something

henrikjaerleblad replied:
In the future, I recommend that you always check the OWCF packages, to see if there is already a package that does what you want :)

anvalen97 replied:
alright! Good with me :)

henrikjaerleblad commented on fusreact.py:
```diff
@@ -178,7 +180,7 @@ def sigma_tot(E, reaction='d-d'):
 
 
-def sigma_diff(E, costheta, reaction='d-d', product_state='GS', anisotropic=True):
+def sigma_diff(E, costheta, reaction='d-d', product_state='gs', anisotropic=True):
```
This is also already fixed in the main branch

henrikjaerleblad commented on fusreact.py:
```diff
@@ -268,6 +270,30 @@ def sigma_diff(E, costheta, reaction='d-d', product_state='GS', anisotropic=True
+        # Interpolate over requested values; zero values will give no contribution to the spectra
+        E_deuterium_beam = E * (constants.md+constants.mt)/constants.mt
+        sigma = interpolatore((E_deuterium_beam,costheta)) /2/np.pi # m**2/sr;
+
```
This addition needs more commenting. More exactly what every line does etc

henrikjaerleblad commented on helper/calcOrbSpec.jl:
```diff
@@ -33,13 +33,14 @@ Calculate the expected diagnostic spectrum of one orbit. The inputs are as follo
 - Ed_bins - The diagnostic measurement bins into which the synthetic measurements will be binned
 - reaction - The reactants of the considered fusion reaction. Format is 'a-b' where a is the thermal species, and b is the fast-ion species
 Keyword arguments include
+- reaction_name - Needed to define the reaction within the analytical calculation routine
```
Again, this either you or I will fix, by extending the `reaction_full` variable

henrikjaerleblad commented on helper/calcOrbSpec.jl:
```diff
@@ -104,12 +105,20 @@ function calcOrbSpec(M::AbstractEquilibrium, o::Orbit{Float64, EPRCoordinate{Flo
-        #!# call LOS file 
-        #!# compute analytical spectra on the intersected points
-        #!# return spectrum
+        include(folderpath_OWCF*"vcone.jl")
```
These include() statements should be written at the top of this script, to avoid having to execute them every time the calcOrbSpec() function is called

anvalen97 replied:
righto

henrikjaerleblad commented on helper/calcOrbSpec.jl:
```diff
@@ -104,12 +105,20 @@ function calcOrbSpec(M::AbstractEquilibrium, o::Orbit{Float64, EPRCoordinate{Flo
     else 
         spec = zeros(length(Ed_bins)-1)
-        stats = 1
+        stats = 10
```
This I have solved now with a new `n_gyro` input variable in the `templates/start_calcOW_template.jl` input file

anvalen97 replied:
I dont really know how to change this honestly. Is it a merge or something?

anvalen97 replied:
P.S. I just mean that we do not really see the proper version here, and I'm not sure what to do

henrikjaerleblad commented on misc/availReacts.jl:
```diff
@@ -102,8 +112,15 @@ function checkReaction(reaction_full::String; verbose::Bool=false, projVelocity:
+
+    # Return the reaction type string only
+    return reaction_name
+end
```
I strongly believe this `reaction_name` variable is not the way to go. Maybe we should discuss this, but since you wrote about it in the pull-request text, I think we should go with the extending-reaction-full approach instead. Since the `9Be(4He,12C)n-1L` approach can contain all the information in one single variable

anvalen97 replied:
I will try to amend here. Feel free to correct this part of the code and others!

henrikjaerleblad commented on spec.py:
```diff
@@ -427,7 +427,7 @@ def __call__(self, bins=None, bin_width=25.0, normalize=False):
         if self.u1 is None:
             u1 = sampler.sample_sphere(self.n_samples)   # random emission directions
-        elif self.product_state == 'GS': #!#
+        elif self.product_state == 'gs': #!#
```
This I also already fixed in the main branch :)

henrikjaerleblad commented on vcone.jl:
```diff
@@ -56,6 +56,15 @@ struct ViewingCone
+    println(io,"minimum(z), maximum(z) [meters]: $(round(minimum(vc.Z),sigdigits=4)), $(round(maximum(vc.Z),sigdigits=4))")
+    println(io,"Resolution (dR,dphi,dz) [meters]: ($(round(vc.dR,sigdigits=4)),$(round(vc.dPHI,sigdigits=4)),$(round(vc.dZ,sigdigits=4)))")
+end
+
```
This is good. Perfect :)

henrikjaerleblad commented on vcone.jl:
```diff
@@ -194,12 +203,12 @@ i_points: array of point indices (same length as i_voxels).
     multiple voxels, since toroidal symmetry is assumed).
 """
-function map_points_voxels(vc::ViewingCone, R::Union{Real,T}, z::Union{Real,T}) where {T<:AbstractArray}
+function map_points_voxels(vc::ViewingCone, R::Union{Real,T}, z::Union{Real,T}; verbose=false, vverbose=false) where {T<:AbstractArray}
```
Also, very nice. Approved! :)

henrikjaerleblad commented on vcone.jl:
```diff
@@ -211,13 +220,17 @@ function map_points_voxels(vc::ViewingCone, R::Union{Real,T}, z::Union{Real,T})
     for (i, n) in enumerate(nvox)
         i1 = i0 + n - 1
+        vverbose && println("map_points_voxels() -> i0: $(i0)    n: $(n)    i1: $(i1)     i: $(i)")
+        vverbose && println("map_points_voxels() -> i_points_inside[i]: $(i_points_inside[i])")
```
Yep, I approve of everything in the vcone.jl script :+1:

henrikjaerleblad reviewed (changes requested):
Alright, so there are quite a lot of changes that need to be made before we can merge this branch with the `main` branch. However, let's team up to implement those changes together. I will start looking at the conflicts, to see what should be done. And maybe @anvalen97 can start going through all my comments?

henrikjaerleblad commented on Project.toml:
```diff
@@ -40,6 +47,8 @@ SCS = "c946c3f1-0d1f-5ce8-9dea-7daa1f7e2d13"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
 StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
+StructArrays = "09ab397b-f2b6-538f-b94a-2f83cf4a842a"
```
One comment regarding how bad I have been myself in determining whether a package is necessary or not for the OWCF: Yesterday, I made a package clean-up where I removed OWCF packages that were no longer used/necessary. I am embarrassed to admit that I removed a total of more than 12 packages. 12! In short, knowing whether a package is necessary or not for the OWCF is not that easy. So no worries if you don't optimize it on your first try @anvalen97 :)

anvalen97 replied:
yes, this one was really installed by accident. Sorry

anvalen97 commented:
Ahaha hej there! I will start now with looking at your comments :)
Slow and steady, I'm sure we'll make it!
I appreciate your patience with going through all the code, line by line.

I can do everything from github, right? Both replying to your comments and tweak the code where needed, no?
Cheers,
Andrea


________________________________
From: Henrik Järleblad ***@***.***>
Sent: 24 January 2025 10:39 AM
To: JuliaFusion/OWCF ***@***.***>
Cc: Andrea Valentini ***@***.***>; Mention ***@***.***>
Subject: Re: [JuliaFusion/OWCF] Implementation of 1step and 2step analytical orbit spectra (PR #66)


@henrikjaerleblad commented on this pull request.

________________________________

In Project.toml<https://github.com/JuliaFusion/OWCF/pull/66#discussion_r1928404281>:

>  StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
+StructArrays = "09ab397b-f2b6-538f-b94a-2f83cf4a842a"


One comment regarding how bad I have been myself in determining whether a package is necessary or not for the OWCF: Yesterday, I made a package clean-up where I removed OWCF packages that were no longer used/necessary. I am embarrassed to admit that I removed a total of more than 12 packages. 12! In short, knowing whether a package is necessary or not for the OWCF is not that easy. So no worries if you don't optimize it on your first try @anvalen97<https://github.com/anvalen97> :)

—
Reply to this email directly, view it on GitHub<https://github.com/JuliaFusion/OWCF/pull/66#discussion_r1928404281>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BLBQZB3O7444NFTISTO5MHT2MIC6JAVCNFSM6AAAAABVVIYFROVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZSGE3DAMJUGY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

anvalen97 commented on Project.toml:
```diff
@@ -21,16 +23,21 @@ Interact = "c601a237-2ae4-5e1e-952c-7a85b0c7eef1"
 Optim = "429524aa-4258-5aef-a3af-852621145aeb"
 OrbitTomography = "9be2d679-58c4-56f3-b3be-0fde37438d3e"
 OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
+PlotlyJS = "f0f68f2c-4968-5e81-91da-67840de0976a"
```
yes, all good with me!
PyPlot I will certainly use it more in the future, so I'll find some way to never commit it by accident every time.

anvalen97 commented on Project.toml:
```diff
@@ -21,16 +23,21 @@ Interact = "c601a237-2ae4-5e1e-952c-7a85b0c7eef1"
 OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
+PlotlyJS = "f0f68f2c-4968-5e81-91da-67840de0976a"
 Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
+Polynomials = "f27b6e38-b328-58d1-80ce-0feddd5e7a45"
```
If we know that that LegendrePolynomials does not depend on this one directly, agreed!

henrikjaerleblad replied:
If it does, it will be included automatically without having to be included explicitly on this list :)

anvalen97 commented on Project.toml:
```diff
@@ -40,6 +47,8 @@ SCS = "c946c3f1-0d1f-5ce8-9dea-7daa1f7e2d13"
 SharedArrays = "1a1011a3-84de-559e-8e89-a11a2f7dc383"
 Sobol = "ed01d8cd-4d21-5b2a-85b4-cc3bdc58bad4"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
```
agreed!

henrikjaerleblad commented on Project.toml:
```diff
@@ -21,16 +23,21 @@ Interact = "c601a237-2ae4-5e1e-952c-7a85b0c7eef1"
 LaTeXStrings = "b964fa9f-0449-5b57-a5c2-d3ea65f4040f"
+LegendrePolynomials = "3db4a2ba-fc88-11e8-3e01-49c72059a882"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+LoopVectorization = "bdcacae8-1622-11e9-2a5c-532679323890"
```
After having looked through the forward.jl script, I agree that the package LoopVectorization is a useful addition both to the script, but also the OWCF in general.

anvalen97 replied:
Yes, it is pretty cool this one. Although admittedly I may not be experienced enough to use it to its full potential.

anvalen97 commented:
[heart]         Andrea Valentini reacted to your message:
________________________________
From: Henrik Järleblad ***@***.***>
Sent: Friday, January 24, 2025 1:05:53 PM
To: JuliaFusion/OWCF ***@***.***>
Cc: Andrea Valentini ***@***.***>; Mention ***@***.***>
Subject: Re: [JuliaFusion/OWCF] Implementation of 1step and 2step analytical orbit spectra (PR #66)


@henrikjaerleblad commented on this pull request.

________________________________

In Project.toml<https://github.com/JuliaFusion/OWCF/pull/66#discussion_r1928652318>:

>  LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+LoopVectorization = "bdcacae8-1622-11e9-2a5c-532679323890"


After having looked through the forward.jl script, I agree that the package LoopVectorization is a useful addition both to the script, but also the OWCF in general.

—
Reply to this email directly, view it on GitHub<https://github.com/JuliaFusion/OWCF/pull/66#discussion_r1928652318>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BLBQZB5XFFUUJRSLHSK6LYL2MI3DDAVCNFSM6AAAAABVVIYFROVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZSGYYDEMBQHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

anvalen97 commented:
> 
> No worries! I appreciate you taking the time to learn Julia and to develop the OWCF further :)
> 
> Yes, in theory you could edit the code directly on Github. You have to start a codespace however. I have not done so yet, because I want to have my local code to always reflect what is going on in the cloud. But if you can start a codespace, please go ahead. Otherwise, if I as an admin need to do it, please let me know !

Hi again! I did not know what it was, but it looks interesting! I have booted up a codespace now, and it seems to be working, It's also quite handy because I can see code and comments altogether. I'll start now with replying (it'll take some time I think)

anvalen97 commented on Project.toml:
```diff
@@ -21,16 +23,21 @@ Interact = "c601a237-2ae4-5e1e-952c-7a85b0c7eef1"
+LoopVectorization = "bdcacae8-1622-11e9-2a5c-532679323890"
 MAT = "23992714-dd62-5051-b70f-ba57cb901cac"
 Mux = "a975b10e-0019-58db-a62f-e48ff68538c9"
+NaNMath = "77ba4419-2d1f-58cd-9bb1-8ffee604a2e3"
```
Let's keep it for the moment. I basically need a marker at some point while calculating the emitted product energies. And having specific elements of the arrays put to NaN simplifies a lot the workflow. I think I can find some other way though: if that's the case, I'll remove this as well

anvalen97 commented on apps/weightsWebApp.jl:
```diff
@@ -884,10 +887,16 @@ function app(req)
+            else
+                pmci = argmin(abs.(pm_array .- pm)) # Find the closest match
+                Rmci = argmin(abs.(Rm_array .- Rm)) # Find the closest match
+                spec = W_correct[:,Ei,pmci,Rmci]./maximum(W_correct[:,Ei,pmci,Rmci])
```
Alright, I'll put it to resolved, and then you can tweak it eventually?

anvalen97 commented on calcOrbWeights.jl:
```diff
@@ -94,11 +94,13 @@ end
 if !analyticalOWs
     reaction_full = deepcopy(reaction) # Make a fully independent copy of the fusion reaction variable
     reaction = full2reactsOnly(reaction; verbose=verbose, projVelocity=analyticalOWs) # Converts from 'a(b,c)d' format to 'a-b' format (reactants only)
+    reaction_name = getReactionName(reaction_full)
```
Regarding this issue, shall I leave to you the changes in calcOrbWeights, availReacts, and CalcOrbSpec? Meanwhile I can change the inputs for forward.jl according to target(beam,detected)residual-nuclearLevel. It's just a proposition, but I think it's the best way we can proceed from here. Lemme know!

I'll paste this wherever pertinent.

anvalen97 commented on calcOrbWeights.jl:
```diff
@@ -494,6 +496,7 @@ F_os = 1.0 .*ones(size(og_orbs)) # Assume one ion for every orbit. This is the d
 # Setting Python variables and structures on all distributed workers/processes...
 verbose && println("Setting all Python variables and structures on all distributed workers/processes... ")
 @everywhere begin
+if analyticCalc
```
yes, your are definitely right.

anvalen97 commented on calcOrbWeights.jl:
```diff
@@ -581,7 +594,7 @@ for iii=1:iiimax
                 @async begin
                     W = @distributed (+) for i=1:norbs
-                        spec = calcOrbSpec(M, og_orbs[i], F_os[i], py"forwardmodel", py"thermal_dist", py"Ed_bin_edges", reaction; product_state=product_state, thermal_temp=thermal_temp, thermal_dens=thermal_dens) # Calculate the diagnostic energy spectrum for the orbit
+                        spec = calcOrbSpec(M, og_orbs[i], F_os[i], py"forwardmodel", py"thermal_dist", py"Ed_bin_edges", reaction, reaction_name; product_state=product_state, thermal_temp=thermal_temp, thermal_dens=thermal_dens,analyticCalc=analyticCalc) # Calculate the diagnostic energy spectrum for the orbit
```
Regarding this issue, shall I leave to you the changes in calcOrbWeights, availReacts, and CalcOrbSpec? Meanwhile I can change the inputs for forward.jl according to target(beam,detected)residual-nuclearLevel. It's just a proposition, but I think it's the best way we can proceed from here. Lemme know!

I'll paste this wherever pertinent.

anvalen97 replied:
P.S. please note that "reaction_name" is an internal variable, that the user would not need to input. It is provided by availReacts.jl.
In any case, let us change the approach here.

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+#
+# Queries for Henrik:
+#  - specifying all the argument types for all functions is not necessary here, since most inputs are defined ¨
+#    right here in this script. This saves on computation time.
```
So, this is what I also thought initially. In fact, that is why you see those commented lines in the beginning of the script, where each function had all input types specified. 
However, this (for reasons I don't understand) made the computation time significantly higher. That does not make sense to me either, but I interpreted it as Julia not having to check the types at all, and so making everything faster; which is crucial here. Do you agree with this?
In any case, type-specific-function (and not) are available already -- so we can decide at any moment to switch from one to the other.

anvalen97 replied:
P.S. I stress again that 90% of variables are internal, and so there would be no way stuff is initialized wrongly (getting unwanted errors, or, worse, no errors and wrong computations). I think we'd be on the safe side, if that makes sense.

anvalen97 replied:
Hi Henrik! Sorry for replying after two months almost, but I completely missed this conversation.
I'm glad we get to agree on this! And that the community also helped.

As the issue seems resolved to me, I simply ask your confirmation that no additional changes are expected here.

anvalen97 replied:
See reply in the previous similar conversation!

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+using LegendrePolynomials  # For Legendre polynomials
+using LinearAlgebra
+
+cnst = pyimport("constants")
```
Yep, I was doing this in the scratch file indeed. Fixed now

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+#                    reaction_name::String="TD_neutrons", 
+#                    bulk_dens::Union{Nothing,Vector{Float64}}=1.0e19)
+#     # ... function body ...
+# end
```
Please look at comment above

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+                min.(-π/2 .+ 1.15 .* acos.(cosLABmin), -0.001),)
+end
+
+function Doppler_spectrum(E_nominal, E_g, phi, n_b, E_b, l_b, m_b, dOmega)
```
see comment above

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+    cos_phi = cos.(phi)
+    sin_phi = sin.(phi)
+
+    sin_gyro = @. (( (E_g/E_nominal)^-1 - G_b + sqrt(G_b2 - 1) * cos_l_b * cos_phi ) / 
```
what do you mean exactly? ahah

anvalen97 replied:
Alright, let me know!

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+                  ( sqrt(G_b2 - 1) * sin_l_b * sin_phi ))
+    cosLAB = @. cos_l_b * cos_phi - sin_l_b * sin_phi * sin_gyro
+
+    dvd_dcosLAB = 0
```
this is to emphasize that in this specific case the emitted product velocity cannot change, since it's a photon.
I think it's a bit more pedagogical this way, but we can also decide to remove it entirely

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+function relativistic_inverse_jacobian(γ_b, cosLAB, E_b, m_b, m_t, m_d, m_r, A, B, D)
+    s = -sign.(cos.(γ_b))
+
+    jac = s .* ( (E_b .* (E_b .+ 2 .* m_b) .* cosLAB .* 
```
I admittely put a few too many dots ahah. I'll try to remove the unecessary ones

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+        omega = 4*np.pi # Spherical emission, in all directions
+        dphi = 2*np.pi # All toroidal angles are included
+    else
+        i_voxels, i_points = map_points_voxels(viewing_cone, o_R, o_z) # Map the R,z sample points to the viewing cone voxels
```
righto

henrikjaerleblad replied:
It's really a matter of Julia code aesthetics and convention. Also, you don't want to do the include statement in the middle of for loops (this code will be inside a big loop in calcOrbWeights.jl). Hence, you put it at the top of forward.jl so that it's done when forward.jl is loaded at the top of calcOrbWeights.jl

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+        p_b_vc = o_p[i_points] # Extract pitch points within diagnostic viewing cone
+        R_vc = o_R[i_points] # Extract R points within diagnostic viewing cone
+        z_vc = o_z[i_points] # Extract z points within diagnostic viewing cone
+        bulk_dens_vc = bulk_dens[i_points]
```
I thought it could only be a Vector at this point of the code. If I'm wrong, then I don't really know how I could fix this properly.

henrikjaerleblad replied:
I think this was related to it being allowed to be Nothing. This is not allowed anymore, and taken care of in calcOrbWeights.jl :)

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+    return costanti .* dSigma_dcosLAB .* jacobian
+end
+
+function SetReaction(reaction_name, product_state="GS")
```
I'll try to change it

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+            deleteat!(p_list, isnan.(w_list))
+            filter!(!isnan, w_list)
+        
+            @turbo for j in eachindex(Ed_vals)
```
thanks! I'd like to discuss this more in the future :)

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+            else 
+        m_b,m_t,m_d,m_r,dSigma_dcosCM = SetReaction(reaction_name)
+
+        for i in eachindex(Ed_vals)
```
for 1 step, the code is already pretty fast. I will work on LoopVectorizing this function as well. Maybe in the next few weeks

anvalen97 commented on forward.py:
```diff
@@ -33,7 +33,7 @@ def __repr__(self):
 
     def calc(self, E, p, R, z, weights, bulk_dist, Ed_bins, B_vec,
-            n_repeat=1, product_state='GS', reaction='d-d', bulk_temp='', bulk_dens='', flr=False, v_rot=np.reshape(np.array([0.0,0.0,0.0]),(3,1)), prompt_gamma_energy=None):
+            n_repeat=1, product_state='GS', reaction='d-d', bulk_temp='', bulk_dens='', flr=False, v_rot=np.reshape(np.array([0.0,0.0,0.0]),(3,1))):
```
is it good then?

anvalen97 reviewed (commented):
I had everything ready in the codespace, and wasn't sure what to do.
Hence I just commit-pushed everything from there. I hope that's ok!

henrikjaerleblad commented on helper/calcOrbSpec.jl:
```diff
@@ -1,45 +1,63 @@
-# forward model/synthetic signal computation code, e.g. FIDASIM.
+# As of this version of the OWCF, the functions of calcOrbSpec.jl can compute expected signals via the following two forward models:
+# - Using Monte Carlo sampling with the DRESS code (J. Eriksson et al, Computer Physics Communications 199 (2016) 40–46)
+# - Using analytic formulas developed by A. Valentini ([REFERENCE TO BE INCLUDED HERE])
```
@anvalen97 could you specify a proper reference to your paper here please?

henrikjaerleblad reviewed (changes requested):
We are getting close to merging the Andrea branch with the main branch! @anvalen97 please address all my comments and I will fix the merge conflicts with the main branch once you are done. Then, finally, we merge Andrea with the main branch, and update the OWCF to v1.5 :)

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,372 @@
+#  - 
+#
+# TO DO:
+#  - check the jacobian for the gamma-ray in spec.py as well. It might not be there
```
I think we can resolve this for now, and leave the comment in the script. This jacobian is not incredibly important for the spectra computations but, yes, it is technically missing. I can fix that at some point, but I would not worry about it since both Massimo and I verified that it leads to minimal modification of the spectral shape.

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,422 @@
+- m_r - mass in keV of the residual particle
+- dSigma_dcosCM - differential cross section in m^2 for specified reaction. NB it is not per unit steradian!
+"""
+function dn_dtdγ(γ_b, l_d, m_d, n_b, E_b, l_b, m_b, n_t, m_t, m_r, dSigma_dcosCM)
```
yes, but this part of the code is not the bootleneck of the computation, by far.
I avoid using the LoopVectorization everywhere but where most needed. I hope that makes sense to you!

anvalen97 commented on forward.jl:
```diff
@@ -0,0 +1,422 @@
+- m_r - mass in keV of the residual particle
+- dSigma_dcosCM - differential cross section in m^2 for specified reaction. NB it is not per unit steradian!
+"""
+function dn_dtdE(E_d::Vector{Float64}, l_d::Vector{Float64}, m_d, n_b, E_b, l_b, m_b, n_t, m_t, m_r, dSigma_dcosCM::Function)
```
We can discuss this if you want, but see previous comment also.
Looking forward to discuss!

henrikjaerleblad replied:
I implemented it myself, since it's mainly a question of code aesthetics then :)

anvalen97 commented on apps/weightsWebApp.jl:
```diff
@@ -869,25 +882,36 @@ function app(req)
 
         # Signal plot
-        if plot_S
+        if plot_S || plot_WF || show_delta_signal # CONTINUE CODING HERE! WEIGHT DELTA SIGNAL WILL NOT BE SAME ORDER OF MAGNITUDE AS S AND WF
```
I dont know, I dont think I wrote that comment honestly

anvalen97 commented on apps/weightsWebApp.jl:
```diff
@@ -147,6 +147,7 @@ diagnostic_name = "" # Diagnostic sightline aestethic keyword. E.g: "TOFOR", "AB
 filepath_W = ""
 (plot_Fos = false) && (filepath_Fos3D = "") # Enable and specify an orbit-space fast-ion distribution file (3D-format)
 (plot_S = false) && (filepath_spec = "")
+filepath_spec = ""
```
this should be outdated now

anvalen97 commented on apps/weightsWebApp.jl:
```diff
@@ -884,10 +887,16 @@ function app(req)
                 plt_sig = Plots.scatter!([Ed_array_WF[Edi_WF]],[(S_WF[irec,:])[Edi_WF]],markersize=5.0,markercolor=sig_color, title="Ed: $(round(Ed,digits=4)) "*Ed_units*"  S: $(round((spec)[Edi_S],sigdigits=4))  WF: $(round((S_WF)[Edi_WF],sigdigits=4))", label="", legend=true)
                 plt_sig = Plots.scatter!([Ed_array_S[Edi_S]],[spec[Edi_S]],markersize=5.0,markercolor=:black, label="")
-            else
+            elseif filepath_spec !== ""
```
this should be outdated now

henrikjaerleblad reviewed (dismissed):
Everything should be ready now to merge the Andrea branch with the main branch. Small bugs will inevitably slip through. If such bugs are encountered, please open an issue. To ensure correct OWCF functionality, safety tests are planned. These will be included soon.

anvalen97 commented:
> Everything should be ready now to merge the Andrea branch with the main branch. Small bugs will inevitably slip through. If such bugs are encountered, please open an issue. To ensure correct OWCF functionality, safety tests are planned. These will be included soon.
Marvellous! I can see there is only two unresolved things from Jan24, but very minor things. If those are taken care of, I think we can go for the merge :)

henrikjaerleblad commented:
> > Marvellous! I can see there is only two unresolved things from Jan24, but very minor things. If those are taken care of, I think we can go for the merge :)

Working on them now. Also removing the 'stats=100' thing in calcOrbSpec when computing 2-step GRS spectra. This is no longer needed, since you will be able to specify the discretization of the gyro motion in the start_calcOW_template.jl file now.
